Make bus queriable (#54)
zenstruck/messenger-test: README.md

                    async: test://
    ```

## Transport

You can interact with the test transports in your tests by using the
`InteractsWithMessenger` trait in your `KernelTestCase`/`WebTestCase` tests.
You can assert the different steps of message processing by asserting on the queue

> and the different state of message processing like acknowledged, rejected and so on.

Reviewer: ```suggestion
and the different states of message processing like "acknowledged", "rejected" and so on.
```

> If you only need to know if a message has been dispatched you can test on the bus. (see below)

Reviewer 2: ```suggestion
> **Note**: If you only need to know if a message has been dispatched you can make assertions [on the bus itself](#bus).
```

    public function test_something(): void
    {
        $this->transport('high')->queue();
        $this->transport('low')->dispatched();
    }
}
```

## Bus

> In addition to transport testing you also can test on the bus. You can test message

Reviewer 2: ```suggestion
In addition to transport testing you also can make assertions on the bus. You can test message
```

use Zenstruck\Messenger\Test\InteractsWithMessenger;

class MyTest extends KernelTestCase
{
    use InteractsWithMessenger;

    public function test_something(): void
    {
        // ... some code that use bus

        // Let's assume two messages are processed
        $this->bus()->dispatched()->assertCount(2);

        $this->bus()->dispatched()->assertContains(\MessageA::class, 1);
        $this->bus()->dispatched()->assertContains(\MessageB::class, 1);
    }
}
```

### Multiple buses

If you use multiple buses you can test that a specific bus has handled its own
messages.

```yaml
# config/packages/messenger.yaml

# ...

framework:
    messenger:
        default_bus: bus_c
        buses:
            bus_a: ~
            bus_b: ~
            bus_c: ~
```

In your tests, pass the name to the `bus()` method:
```php
use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Zenstruck\Messenger\Test\InteractsWithMessenger;

class MyTest extends KernelTestCase
{
    use InteractsWithMessenger;

    public function test_something(): void
    {
        // ... some code that use bus

        // Let's assume two messages are handled by two different buses

>         $this->bus('bus-a.test-bus')->dispatched()->assertCount(1);
>         $this->bus('bus-b.test-bus')->dispatched()->assertCount(1);
>         $this->bus('bus-c.test-bus')->dispatched()->assertCount(0);
> 
>         $this->bus('bus-a.test-bus')->dispatched()->assertContains(\MessageA::class, 1);
>         $this->bus('bus-b.test-bus')->dispatched()->assertContains(\MessageB::class, 1);

Reviewer 2: ```suggestion
        $this->bus('bus-a')->dispatched()->assertCount(1);
        $this->bus('bus-b')->dispatched()->assertCount(1);
        $this->bus('bus-c')->dispatched()->assertCount(0);

        $this->bus('bus-a')->dispatched()->assertContains(\MessageA::class, 1);
        $this->bus('bus-b')->dispatched()->assertContains(\MessageB::class, 1);
```

_Once this is possible._

    public function test_something(): void
    {
        $this->transport('high')->queue();
        $this->transport('low')->dispatched();
    }
}
```

## Bus

In addition to transport testing you also can make assertions on the bus. You can test message
handling by using the same `InteractsWithMessenger` trait in your `KernelTestCase` / `WebTestCase` tests.
This is especially useful when you only need to test if a message has been dispatched
by a specific bus but don't need to know how the handling has been made.

It allows you to use your custom transport while asserting your messages are still dispatched properly.

> ### Single bus
> ```php

Reviewer 2: ```suggestion
### Single bus

```php
```

use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Zenstruck\Messenger\Test\InteractsWithMessenger;

class MyTest extends KernelTestCase
{
    use InteractsWithMessenger;

    public function test_something(): void
    {

>         // ... some code that use bus

Reviewer: ```suggestion
        // ... some code that uses bus
```

Reviewer 2: ```suggestion
        // ... some code that uses the bus
```

        // Let's assume two messages are processed
        $this->bus()->dispatched()->assertCount(2);

>         $this->bus()->dispatched()->assertContains(\MessageA::class, 1);
>         $this->bus()->dispatched()->assertContains(\MessageB::class, 1);

Reviewer 2: ```suggestion
        $this->bus()->dispatched()->assertContains(MessageA::class, 1);
        $this->bus()->dispatched()->assertContains(MessageB::class, 1);
```

    }
}
```

### Multiple buses

If you use multiple buses you can test that a specific bus has handled its own
messages.

```yaml
# config/packages/messenger.yaml

# ...

framework:
    messenger:
        default_bus: bus_c
        buses:
            bus_a: ~
            bus_b: ~
            bus_c: ~
```

> In your tests, pass the name to the `bus()` method:
> ```php

Reviewer 2: ```suggestion
In your tests, pass the name to the `bus()` method:

```php
```

use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Zenstruck\Messenger\Test\InteractsWithMessenger;

class MyTest extends KernelTestCase
{
    use InteractsWithMessenger;

    public function test_something(): void
    {
        // ... some code that use bus

        // Let's assume two messages are handled by two different buses
        $this->bus('bus-a')->dispatched()->assertCount(1);
        $this->bus('bus-b')->dispatched()->assertCount(1);
        $this->bus('bus-c')->dispatched()->assertCount(0);

>         $this->bus('bus-a')->dispatched()->assertContains(\MessageA::class, 1);
>         $this->bus('bus-b')->dispatched()->assertContains(\MessageB::class, 1);

Reviewer 2: ```suggestion
        $this->bus('bus-a')->dispatched()->assertContains(MessageA::class, 1);
        $this->bus('bus-b')->dispatched()->assertContains(MessageB::class, 1);
```
